vendor/github.com/bitrise-io/got/download.go

```diff
@@ -313,10 +325,58 @@ func (d *Download) dl(dest io.WriterAt, errC chan error) {
 		go func(i int) {
 			defer wg.Done()
 
-			// Concurrently download and write chunk
-			if err := d.DownloadChunk(d.chunks[i], &OffsetWriter{dest, int64(d.chunks[i].Start)}); err != nil {
+			// This OffsetWriter allows two things:
+			// - write to the offset of the file
+			// - in case of an interrupt and re-download, it will resume from the last position
+			offsetWriter := &OffsetWriter{dest, int64(d.chunks[i].Start)}
+
+			err := retry.Times(uint(d.MaxRetryPerChunk)).Try(func(attempt uint) error {
+				log := func(msg string, args ...interface{}) {
+					if d.Logger == nil {
+						return
+					}
+					prefix := fmt.Sprintf("[chunk=%d][attempt=%d]%s ", i, attempt, stats.String())
+					d.Logger.Debugf(prefix+msg, args...)
+				}
+
+				// Concurrently download and write chunk
+				start := time.Now()
+
+				// Per-chunk cancellation signal
+				chunkCtx, cancelChunk := context.WithCancel(d.ctx)
+				defer cancelChunk()
+
+				// Check for hanged downloads and interrupt them
+				downloadCheckTicker := time.NewTicker(time.Second)
+				defer downloadCheckTicker.Stop()
+
+				go func() {
+					if attempt == uint(d.MaxRetryPerChunk) {
+						log("last attempt, won't start ticker")
+						return // never interrupt the last try
+					}
+					log("start ticker")
+					for range downloadCheckTicker.C {
+						if stats.finishedChunks > 0 && time.Since(start)-stats.average() > d.ChunkRetryThreshold {
```

Comment:
stats.finishedChunks > 0 may need to lock here to see consistent finishedChunks.